src/Symfony/Component/Security/Http/Impersonate/ImpersonateUrlGenerator.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\Impersonate;
+
+use Symfony\Bundle\SecurityBundle\Security\FirewallMap;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Http\Firewall\SwitchUserListener;
+
+/**
+ * Provides generator functions for the impersonate url exit.
+ *
+ * @author Amrouche Hamza <hamza.simperfit@gmail.com>
+ * @author Damien Fayet <damienf1521@gmail.com>
+ */
+class ImpersonateUrlGenerator
+{
+    private $requestStack;
+
+    private $tokenStorage;
+
+    private $firewallMap;
+
+    public function __construct(RequestStack $requestStack, FirewallMap $firewallMap, TokenStorageInterface $tokenStorage)
+    {
+        $this->requestStack = $requestStack;
+        $this->tokenStorage = $tokenStorage;
+        $this->firewallMap = $firewallMap;
+    }
+
+    public function generateImpersonateExitUrl(string $exitTo = null, $referenceType = UrlGeneratorInterface::ABSOLUTE_URL): string
+    {
+        $request = $this->requestStack->getCurrentRequest();
+
+        if (!$this->isImpersonatedUser()) {
+            return '';
+        }
+
+        if (null === $switchUserConfig = $this->firewallMap->getFirewallConfig($request)->getSwitchUser()) {
+            throw new \LogicException('Unable to generate the impersonate exit URL without a firewall configured for the user switch.');
+        }
+
+        if ('_use_impersonated_from_url' === $exitTo) {
```

Comment:
perhaps we should decouple this into its own (twig) method so we can do

```
{{ impersonation_exit_url(impersonation_start_url()) }}
```

meaning we can implement `impersonation_start_url` as a hybrid;
if impersonated: points to "originated from uri"
else: points to "switch user uri"

Reply:
> and perhaps for completeness `is_impersonated()`

I don't think it's worth adding this function, as this information can be simply retrieved by doing `{% if is_granted('ROLE_PREVIOUS_ADMIN') %}`, your thoughts?

Reply:
About your idea of `impersonation_start_url` should we also consider to create function to get url/path to start impersonation? 

I thing the name `impersonation_start_url` might be misleading as it looks like the `*_exit*` ones, but does not really generate urls to "start" the impersonation.

I would be more confident with names like `impersonation_get_start_url` or `impersonation_started_url`.

Reply:
I'm not sure I like this `_use_impersonated_from_url` feature. I would drop it completely (at least at first).